hack/externalnode/install-vm.sh

```diff
@@ -141,17 +153,278 @@ EOF
+        fi
+    done
+}
+
+start_containers() {
+    while [ $COUNT -lt $LIMIT ] ; do
+        for service in "${CONTAINERS[@]}"; do
+            if [ ! "$(docker ps -q -f name=$service)" ]; then
+                if [ "$(docker ps -aq -f status=exited -f name=$service)" ] ; then
+                    echo "$service container is in exited state"
+                    ((COUNT=COUNT+1))
+                fi
+                echo "Starting containers ${CONTAINERS[*]}"
+                docker-compose -f $DOCKER_COMPOSE_YAML up -d
+            fi
+        done
+        sleep 5
+    done
+}
+
+# To validate any other instance of this script is running in this machine.
+if [ -f "$PIDFILE" ]; then
+    # Kills the another instance of this script process.
+    kill -0 "$(cat $PIDFILE)" && kill -9 "$(cat $PIDFILE)"
+fi
+
+# Remove the pid file which contains this script process id.
+trap remove_pidfile EXIT
+# Stores the current pid of this script.
+echo $$ > "$PIDFILE"
+
+docker info
+if [[ $? -ne 0 ]]; then
+    echo "Starting docker service"
+    systemctl start docker
+    sleep 2
+fi
+
+key=$1
+case $key in
+    start)
+    load_ovs_modules
+    start_containers
+    ;;
+    stop)
+    docker-compose -f $DOCKER_COMPOSE_YAML stop ${CONTAINERS[*]}
+    ;;
+esac
+EOF
+}
+
+# Create antrea-agent service as a systemd service for monitoring/creating antrea specific containers.
+setup_antrea_agent_service() {
+    DESCRIPTION="Antrea agent service to monitor OVS and antrea-agent containers"
+    SERVICE_NAME="antrea-agent.service"
+    IS_ACTIVE=$(sudo systemctl is-active $SERVICE_NAME)
+    create_antrea_agent_service_file
+    sudo chmod +x $AGENT_CONF_PATH/antrea-agent.sh
+    if [ "$IS_ACTIVE" == "active" ]; then
+        # restart the service
+        echo "Systemd service $SERVICE_NAME is running"
+        echo "Restarting $SERVICE_NAME service"
+        sudo systemctl restart $SERVICE_NAME
+    else
+        # create service file
+        echo "Creating systemd service file for $SERVICE_NAME"
+        sudo cat > /etc/systemd/system/${SERVICE_NAME} << EOF
+[Unit]
+Description=$DESCRIPTION
+After=network.target
+[Service]
+ExecStart=$AGENT_CONF_PATH/antrea-agent.sh start
+ExecStop=$AGENT_CONF_PATH/antrea-agent.sh stop
+[Install]
+WantedBy=multi-user.target
+EOF
+        # restart daemon, enable and start service.
+        echo "Reloading daemon and enabling service $SERVICE_NAME"
+        sudo systemctl daemon-reload
+        sudo systemctl enable ${SERVICE_NAME//'.service'/} # remove the extension.
+        sudo systemctl start ${SERVICE_NAME//'.service'/}
+        if [ $? -eq 0 ]; then
+            echo "Antrea-agent service started successfully"
+        else
+            echoerr "Error Antrea-agent service could not be started, rc $status"
+            exit 1
+        fi
+    fi
+}
+
+# Function to deploy and run antrea specific containers.
+deploy_antrea_containers() {
+    set +eo pipefail
+    echo "Downloading docker image $ANTREA_DOCKER_IMAGE"
```

Comment:
Here the script only download antrea-ubuntu images, but in the instruction you mentioned it is supported to build the image from source code, how does the script leverage the manually built images?

Reply:
Maybe you should think about use " docker load -i $localImage" in the script to support manually built images from source code?

Reply:
Actually the statement is wrong. It is "Load docker image". The user copies the image and loads the image with the name "projects.registry.vmware.com/antrea/antrea-ubuntu:latest" so the script automatically utilizes the user built image.
We should not add the " docker load -i $localImage" in script because the user itself manually loads the docker image.